internal/controllers/synthesis/slicecleanup.go

```diff
@@ -81,17 +69,83 @@ func (c *sliceCleanupController) Reconcile(ctx context.Context, req ctrl.Request
-	if doNotDelete {
+	if decision.DoNotDelete {
 		return ctrl.Result{}, nil
 	}
+
 	if err := c.client.Delete(ctx, slice); err != nil {
 		return ctrl.Result{}, fmt.Errorf("deleting resource slice: %w", err)
 	}
 	logger.V(0).Info("deleted unused resource slice")
 	return ctrl.Result{}, nil
 }
 
+func (c *sliceCleanupController) buildCleanupDecision(ctx context.Context, slice *apiv1.ResourceSlice, owner *metav1.OwnerReference) (cleanupDecision, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+	if owner == nil {
+		logger.V(1).Info("resource slice can be deleted because it does not have an owner")
+		return cleanupDecision{}, nil // delete
+	}
+
+	// Bail out early if the cache suggests that we shouldn't delete the resource slice
+	informerDecision, err := checkCompositionState(ctx, c.client, slice, owner)
+	if err != nil {
+		return cleanupDecision{}, err
+	}
+	if informerDecision.DoNotDelete && informerDecision.HoldFinalizer {
+		return informerDecision, nil
+	}
+
+	// Don't run the actual (non-cached) check if the resource is too new - the cache is probably just stale
+	// This is not intended to improve safety, only to reduce the number of unnecessary gets to apiserver
+	age := time.Since(slice.CreationTimestamp.Time)
+	const threshold = time.Second * 5
+	if age < threshold {
+		logger.V(1).Info("refusing to delete resource slice because it's too new", "age", age.Milliseconds())
+		return cleanupDecision{DeferBy: ptr.To(threshold - age)}, nil
+	}
+
+	// Check the state against apiserver without any caching before making a final decision
+	apiDecision, err := checkCompositionState(ctx, c.noCacheReader, slice, owner)
+	if err != nil {
+		return cleanupDecision{}, err
+	}
+	if !reflect.DeepEqual(informerDecision, apiDecision) {
+		// We trust the apiserver decision even if it doesn't align with the cache,
+		// although this is should rarely happen and might be reason for concern
+		logger.Info("cleanup decisions derived from informer cache and non-caching client do not agree!", "cache", informerDecision.String(), "noCache", apiDecision.String())
+	}
+
+	return apiDecision, nil
+}
+
+func checkCompositionState(ctx context.Context, reader client.Reader, slice *apiv1.ResourceSlice, owner *metav1.OwnerReference) (cleanupDecision, error) {
+	comp := &apiv1.Composition{}
+	comp.Name = owner.Name
+	comp.Namespace = slice.Namespace
+	err := reader.Get(ctx, client.ObjectKeyFromObject(comp), comp)
+	if errors.IsNotFound(err) {
+		return cleanupDecision{}, nil // delete
+	}
+	if err != nil {
+		return cleanupDecision{}, fmt.Errorf("getting composition: %w", err)
+	}
+	return cleanupDecision{
+		DoNotDelete:   !shouldDeleteSlice(comp, slice),
+		HoldFinalizer: !shouldReleaseSliceFinalizer(comp, slice),
+	}, nil
+}
+
+type cleanupDecision struct {
+	DoNotDelete   bool
+	HoldFinalizer bool
+	DeferBy       *time.Duration
+}
+
+func (c *cleanupDecision) String() string {
+	return fmt.Sprintf("DoNotDelete=%t,HoldFinalizer=%t", c.DoNotDelete, c.HoldFinalizer)
```

Comment:
No defer by?

Reply:
Yeah, pragmatically is not relevant, just odd to see a field missing without knowing the whole context, will leave it up to you